src/server/transaction/server/driver.go

```diff
@@ -241,44 +247,60 @@ func (d *driver) appendTransaction(
 				return err
 			}
 
-			// Save the length so that we can check that nothing else modifies the
-			// transaction in the meantime
-			numRequests = len(info.Requests)
-			numResponses = len(info.Responses)
 			info.Requests = append(info.Requests, items...)
-			if newResponses != nil {
-				info.Responses = newResponses
+			if newInfo.Responses != nil {
+				info.Responses = newInfo.Responses
 			}
 
 			// This may error with a postgres-level transaction conflict, in which
 			// case we want to save the generated responses and reattempt.  The saved
 			// responses will contain any references to objects generated outside of
 			// the transaction and need to be reused on the next attempt.
 			info, err = d.runTransaction(txnCtx, info)
```

Comment:
I think this function needs the same handling as `FinishTransaction` to map a collection error to a `transactionModifiedError` (for example if the collection transaction error occurs when doing the dryrun when appending a CreatePipeline to an existing transaction) so maybe that logic can just be moved into `updateTransaction`?

Reply:
Hmm, I think the old logic is correct as it was, since it is careful to make sure the SQL transaction is retried with the updated `TransactionInfo`. Regardless, I moved things around so it doesn't have to do this dance.